yoga/Yoga.cpp

```diff
@@ -240,6 +240,11 @@ YGNodeRef YGNodeClone(YGNodeRef oldNode) {
       "Could not allocate memory for node");
   gNodeInstanceCount++;
   node->setOwner(nullptr);
+
+  YGNodeRef root = node;
+  for (; root->getParent() != nullptr; root = root->getParent());
```

Comment:
Looks like this loop always terminates early: the parent is set to `nullptr` above the loop on line 242.

Might be a good idea to add a test? ;-)

Reply:
Oh my apologies, thought it just a simple loop, so didn't test it thoroughly :-(
Okay, let me add a test :+1:

Reply:
No worries.

Thinking about this a bit more, I'm not sure we need to do this extra work on clone: the generation is incremented each time we perform a `YGNodeCalculateLayout`, so even if we preserved the root generation counter for a cloned subtree, we're still incrementing it on a call to `YGNodeCalculateLayout` for that subtree and before we proceed with any work.  Any comparison to the old generation count will be false.

Is the counter used to help avoid node recalculations **during** multiple passes of a given layout (generation) calculation?  I'd need to spend some time to really understand how the generation counter is used for this; and unfortunately the code isn't terribly straightforward or self-documenting.  If you're on the FB team, can you ask the original author for a short-cut?

Reply:
Unfortunately we aren't from the FB team and neither do we have a full understanding how generetionCount works. My team encounter the race condition problem several months ago, so we opened the previous pull request to try to fix that.  
I actually thought you are the one on the FB team ;-)

Reply:
> I would suggest us to stick to the original idea of #791 because it focus on solving the data race condition issue without worrying about optimizing the `generationCount` itself. This can be done in a separated PR and leave this PR with minimal changes.

That would be unwise because #791:
1. uses more heap than necessary;
1. introduces an expensive subtree traversal to set up the pointer-to-the-root, which is performed often; and
1. does not address node cloning.

I am reading through the code to understand how the generation is treated and will report back shortly.

Reply:
Here's what I've learned from reading the code...

There are two generation counters: `generationCount` and `computedFlexBasisGeneration` stored in `YGLayout`, which are compared to the global `gCurrentGenerationCount`.  An overview of each is given below.

At the end, I'll provide a few options on how to move forward with this PR.  Apologies for the length in advance.

#### generationCount

1. This counter is used to invalidate the layout cache in `YGLayoutNodeInternal` when a node is dirty, and the node has not been visited thus far for a given `YGNodeCalculateLayout` call.
1. It is updated on exit of the function whenever layout is performed.
1. In `YGNodeCalculateLayout`, the `gCurrentGenerationCount` is incremented a second time before performing the optional a legacy "stretch behavior" comparison on a cloned tree that has all nodes marked dirty.  The new `gCurrentGenerationCount` here is to ensure the layout cache is cleared as above.
1. It looks like the purpose of this counter is to ensure the layout cache is invalidated whenever a node is dirty once per `YGNodeCalculateLayout` call.


#### computedFlexBasisGeneration

1. This counter is used exclusively in `YGNodeComputeFlexBasisForChild`.
1. It gates computation of a child's computed flex basis when the `YGExperimentalFeatureWebFlexBasis` feature is in use.
1. It is updated on exit of the function every time.
1. It looks like the purpose of this counter is to limit the above computation once per `YGNodeCalculateLayout` call.

#### gCurrentGenerationCount

1. This counter is incremented in `YGNodeCalculateLayout` before performing a full-tree layout.
1. It looks like the purpose of this counter is to uniquely identify a given layout run, which makes sense when you consider its name: "generation" implying that each layout gives birth to a new tree.
1. Because of the way the counter is used (compare with equality only; node update on exit of relevant function), I don't think it matters when it wraps.  This means we could probably get away with an even smaller storage type (e.g. `uint8_t`) for the global counter and associated class member storage.
1. With the above in mind, the comments in the source now make more sense (to me):
    ```
    // Instead of recomputing the entire layout every single time, we
    // cache some information to break early when nothing changed
    ```
    and
    ```
    // Increment the generation count. This will force the recursive routine to
    // visit
    // all dirty nodes at least once. Subsequent visits will be skipped if the
    // input
    // parameters don't change.
    ```

### Moving forward

Some options:

1. Minimize changes:
    * Drop the `YGNodeClone` change.
    * Remove the `generationCount_` member from `YGNode` and bring back the global `gCurrentGenerationCount`.
    * Increment and retrieve `gCurrentGenerationCount` atomically in `YGNodeCalculateLayout` and use it to seed the `generationCount` stack variable that is passed around by function call.
    * Keep the other changes in the current PR, relating to `depth` and `PRINT_XXX`.
1. Minimize heap memory use:
    * Drop the `gCurrentGenerationCount` global.
    * Drop the `computedFlexBasisGeneration` and `generationCount` members of `YGLayout`.
    * Declare a new `std::map<YGNode,uint8_t>` stack variable in `YGNodeCalculateLayout` that is passed around like `depth` is in the current PR.  (Or to minimize function call overhead, the `depth` and a pointer to the map can be placed in a struct, and a pointer to that struct is passed around instead.)
    * The `uint8_t` value field of the map is a bitfield:
        * reserve bit 0 for "I have visited the node" in `YGLayoutNodeInternal` in lieu of the `layout->generationCount != gCurrentGenerationCount` check that currently exists.
        * reserve bit 1 for "I have computed the flex basis for this child" in lieu of the `computedFlexBasisGeneration != gCurrentGenerationCount` check in `YGNodeComputeFlexBasisForChild`.
    * Throw out the map on return of `YGNodeCalculateLayout`.

What do you think?

Reply:
The implementation for the first and second options would be quite straightforward, but I'm having a hard time understanding the third optimization. How exactly can we substitute those 3 variables with the map approach? Personally, I find it really risky to make such a big change, and we don't really have the confidence in our understanding of Yoga to carry out the implementation by ourselves neither.

Reply:
I've submitted PR #855 that takes the third approach: it appears to work well!

So from here, let's see if @davidaurelio or others on the FB team can offer their input.  @tcdat96, in the meantime if you have time to update your PR with option #2, it can give others an opportunity to review both PRs and weigh in.

Hopefully we can land this thread-safety fix soon.  I'm happy to amend my PR if there are preferences on style, etc.

Reply:
I updated the PR with the first 2 options, but ran into error ```undefined reference to `pthread_getspecific'``` when running gtest. Have you ever encountered a similar problem, @jpap?

Reply:
> error `` undefined reference to `pthread_getspecific' `` when running gtest.

Sadly I've never experienced that; sounds like a linker configuration problem?  Perhaps you've inadvertently modified some of the build configuration someplace.  I would try:

* Clean and rebuild?
* Create a new local branch based off master, then cherry-pick **only** your changes then rebuild/test.

Reply:
I've already tried both but unfortunately nothing works. The error appeared even with the master commit so I think it should be my configuration problem, but sadly I don't really have any experience in this type of problem :-(

Reply:
It seems that the error only emerges in our ubuntu machines, the same source code runs without any error in MacOS. 
I modified the BUCK file in lib/gtest with the workaround described in https://github.com/facebook/buck/issues/1443 and it worked. Now this PR runs successfully with all tests passed.